### PR TITLE
Add codegen (contracts, subgraph), order signing and order book API interface

### PR DESCRIPTION
**Pull request by ribeirojose**
Sorry for the gigantic PR 🤠 🐄 😬  

## Contents for this PR:

- Codegen:
    - Create module to generate objects to interact with deployed contracts.
    - Generate Composable CoW, Fallback Handler, Milkman and TWAP Contracts.
- Common:
    - Create API base class with backoff and rate limit configuration
    - Refactor chain and config contants
- Contracts:
    - Add ability to sign and normalize orders with EIP 712, ETH Sign, EIP 1271, and Pre Sign.
- Order Book API:
    - Add models with codegen;
    - Create interface to call public endpoints;
- Subgraph:
    - Add subgraph codegen, queries and clients.
- Examples:
    - Add example of order quoting, signing and posting using OrderBookAPI and Contract Signing.

### CODE REVIEW

File: cow_py/contracts/order.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+#   Order,
+#   "validTo" | "appData" | "kind" | "sellTokenBalance" | "buyTokenBalance"
+# > & {
+#   receiver: string;
+#   validTo: number;
+#   appData: string;
+#   kind: "sell" | "buy";
+#   sellTokenBalance: "erc20" | "external" | "internal";
+#   buyTokenBalance: "erc20" | "internal";
+# };
```
Comment (ribeirojose) [inline cow_py/contracts/order.py:190]: JS code 😛

File: cow_py/contracts/order.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+    # The timestamp this order is valid until.
+    validTo: int
+
+
+# /**
+#  * Computes the order UID for an order and the given owner.
+#  */
+# export function computeOrderUid(
+#   domain: TypedDataDomain,
+#   order: Order,
```
Comment (ribeirojose) [inline cow_py/contracts/order.py:291]: also some JS code here 👀

File: examples/order_posting_e2e.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        }
+    )
+
+    signature = sign_order(order)
+    order_uid = await post_order(order, signature)
+    print(f"order posted on link: {ORDER_BOOK_API.get_order_link(order_uid)}")
+
+
+if __name__ == "__main__":
+    asyncio.run(main())
```
Comment (ribeirojose) [inline examples/order_posting_e2e.py:120]: this still functional after all those refactors @yvesfracari ?

File: tests/codegen/components/test_base_contract.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# tests/codegen/components/test_base_contract.py
```
Comment (ribeirojose) [inline tests/codegen/components/test_base_contract.py:1]: tests missing here

File: tests/codegen/components/test_contract_factory.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+# tests/codegen/components/test_contract_factory.py
```
Comment (ribeirojose) [inline tests/codegen/components/test_contract_factory.py:1]: tests missing here

File: tests/codegen/test_abi_handler.py [hunk 6]
```diff
@@ -16,3 +17,48 @@
 )
 def test_to_python_conventional_name(input_name, expected_output):
     assert to_python_conventional_name(input_name) == expected_output
+
+
+def test_compile_partial():
```
Comment (ribeirojose) [inline tests/codegen/test_abi_handler.py:22]: tests missing here

File: tests/contracts/test_sign.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
+import pytest
+from eth_account.messages import SignableMessage
+from eth_account.signers.local import LocalAccount
+from eth_utils.conversions import to_hex
+from web3 import EthereumTesterProvider, Web3
+
+from cow_py.contracts.order import hash_order_cancellation
+
+# Assuming you have these Python equivalents of your TypeScript code
```
Comment (ribeirojose) [inline tests/contracts/test_sign.py:9]: chatgpt comment 😅

File: tests/contracts/test_sign.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+from cow_py.contracts.sign import SigningScheme, sign_order, sign_order_cancellation
+
+from .conftest import SAMPLE_ORDER
+
+# Setup Web3 and Ethereum Tester (or use your existing setup)
```
Comment (ribeirojose) [inline tests/contracts/test_sign.py:14]: unneeded

File: examples/order_posting_e2e.py (hunk 3 above)
Comment (yvesfracari) [reply examples/order_posting_e2e.py:120]: It should be, I tested yesterday. Did you have any problem running it?

**Comment by mfw78** [comment]
Hey, so yeah, this is gigantic :laughing: I'm going to go out on a limb and estimate that a lot of this code is auto-generated. Can this potentially please be broken into two PRs, one that commits what is needed to do the generation, the other being the auto-generated files (so then reviewers can also determine the success of auto-generation, and SDK-specific logic?) :pray:

**Comment by ribeirojose** [comment]
Closing in favor of breaking down into small PRs:

https://github.com/bleu-fi/cow-py/pull/20
https://github.com/bleu-fi/cow-py/pull/19
https://github.com/bleu-fi/cow-py/pull/18
https://github.com/bleu-fi/cow-py/pull/17
https://github.com/bleu-fi/cow-py/pull/15
https://github.com/cowdao-grants/cow-py/pull/8

@mfw78 to make this easier to review I chained one PR to another's branch, but unfortunately that means that I couldn't open all PRs in this repo since I don't have write permissions here. IMO it's not an issue if you review them in our repo and we work on it there, but lmk if you prefer an alternative to this workflow.

**Pull request closed without merging**